### ISSUE TITLE
[QUESTION] Chapter 13 DATA API: question 10 part e

### ISSUE THREAD

**Issue by kasrahabib** [ageron/handson-ml2#426]
Hi @ageron!

I have problem with understanding the following helper function available for creating an Embedding layer.
This is not a technical issue or that something is not working.

``` python
def compute_mean_embedding(inputs):
    not_pad = tf.math.count_nonzero(inputs, axis=-1) 
    n_words = tf.math.count_nonzero(not_pad, axis=-1, keepdims=True)    
    sqrt_n_words = tf.math.sqrt(tf.cast(n_words, tf.float32))
    return tf.reduce_mean(inputs, axis=1) * sqrt_n_words

```

I understand second line of the code```not_pad``` variable, it counts the number of the non-zero elements and returns it. I do not understand the value in the variable ```n_words```: is it going to provide us with the number of the words in the whole ```training_set``` or only the number words contain each instance.

 For example, I first created a bag of word representation of the ```train_set``` (with the help of the codes available in the note book) as follow:

``` python
x_train = train_set.map(lambda review, label: review)
x_train = np.concatenate(list(x_train), axis = 0)

x_train_vectorized = text_vectorization(x_train)
x_train_bgw = bag_of_words(x_train_vectorized)
x_train_bgw


OUTPUT>>>

                <tf.Tensor: shape=(25000, 1100), dtype=float32, numpy=
                array([[2., 3., 1., ..., 0., 0., 0.],
                       [1., 4., 1., ..., 0., 0., 0.],
                       [1., 0., 0., ..., 0., 0., 0.],
                       ...,
                       [1., 3., 1., ..., 0., 0., 0.],
                       [1., 1., 0., ..., 0., 0., 0.],
                       [1., 1., 1., ..., 1., 0., 0.]], dtype=float32)>
```

Then again tried your code to count the non-zero elements in the rain set:

``` python 
not_pat = tf.math.count_nonzero(x_train_bgw, axis = -1)
not_pat


OUTPUT>>>

<tf.Tensor: shape=(25000,), dtype=int64, numpy=array([40, 42, 44, ..., 40, 44, 44])>
```

Up to this point, I see that the first review contains 40 word, the second 42 and so on. Now here is my question with the following line of the code:

``` python
n_words = tf.math.count_nonzero(not_pat, axis = -1, keepdims = True)
n_words


OUTPUT>>>

<tf.Tensor: shape=(1,), dtype=int64, numpy=array([25000])>
```
I get ```[25000]``` which is the size of the previous variable ```not_pad```. What I was expecting was to be the sum of the whole words in the trappings set something maybe like: ``` tf.reduce_sum(not_pat)``` which gives the number of all available words in the ```train_set```.

Or maybe I am getting it wrong, since during the training ```compute_mean_embedings``` is executed once for each review and each review is multiplied by the square root of its own number of words: for example the first review is multiplied by ```sqrt(40)``` and the second ```sqrt(42)``` and so on.

I appreciate your passion clarifying this question: ☺️

**Comment by ageron** [comment]
Hi @kasrahabib ,

Thanks for your question! I propose to use a simpler input to understand what's happening:

```python
>>> inputs = tf.constant([[[1., 2., 3.], [4., 5., 0.], [0., 0., 0.]],
...                       [[6., 0., 0.], [0., 0., 0.], [0., 0., 0.]]])
...
```

Note that this is a 3D array, not a 2D array like your `x_train_bgw`. Indeed, after the embedding layer, each word is represented as a vector, so a sentence is represented as a 2D array, and a list of padded sentences is represented as a 3D array. I think this is probably the source of the confusion.

The above `inputs` array represents two very short sentences: the first one has two words (embedded as [1., 2., 3.] and [4., 5., 0.] respectively), and the second has just one word (embedded as [6., 0., 0.]). Both sentences are padded with padding tokens (represented as a zero vector: [0., 0., 0.]). The first sentence is padded with 1 padding token, and the second sentence is padded with 2 padding tokens.

The first time we use `count_nonzero()`, we're just counting the number of non-zero values in each token's embedding:

```python
>>> not_pad = tf.math.count_nonzero(inputs, axis=-1)
>>> not_pad
<tf.Tensor: shape=(2, 3), dtype=int64, numpy=
array([[3, 2, 0],
       [1, 0, 0]])>
```

Indeed, the first sentence has three tokens: the first token's embedding has 3 non-zero values (1., 2. and 3.), the second has 2 (4. and 5.), and the last token is the padding token, full of zeros (so there are 0 non-zero values).
The second sentence also has three tokens: the first one's embedding has just 1 non-zero value (6.), and the other two tokens are padding tokens, so there are 0 non-zero values.

With this we're ready to count the number of words in each sentence. To do this, we count the number of tokens whose embeddings have at least one non-zero value. For the first sentence, `not_pad[0]` is `[3, 2, 0]`, which contains 2 non-zero values: this tells us that there are two words in the first sentence (we're simply ignoring the 3rd token, which is a padding token and has 0 non-zero values, since it is represented as [0., 0., 0.]). In the second sentence, `not_pad[1]` is `[1, 0, 0]` which contains just 1 non-zero value. So the second sentence has just 1 word (the rest are padding tokens).

```python
>>> n_words = tf.math.count_nonzero(not_pad, axis=-1, keepdims=True)
>>> n_words
<tf.Tensor: shape=(2, 1), dtype=int64, numpy=
array([[2],
       [1]])>
```

Next, let's compute the square root of the number of words. Since `n_words` has an integer type, we need to cast it to `float32` first:

```python
>>> sqrt_n_words = tf.math.sqrt(tf.cast(n_words, tf.float32))
>>> sqrt_n_words
<tf.Tensor: shape=(2, 1), dtype=float32, numpy=
array([[1.4142135],
       [1.       ]], dtype=float32)>
```

Now let's compute the mean of the embeddings, over all the words in each sentence:

```
>>> mean = tf.reduce_mean(inputs, axis=1)
>>> mean
<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
array([[1.6666666, 2.3333333, 1.       ],
       [2.       , 0.       , 0.       ]], dtype=float32)>
```

The mean embedding in the first sentence is `[1.6666666, 2.3333333, 1.]`, and the mean embedding in the second sentence is `[2., 0., 0.]`.

Lastly, the function returns the mean embedding multiplied by the number of words, for each sentence:

```python
>>> mean * sqrt_n_words
<tf.Tensor: shape=(2, 3), dtype=float32, numpy=
array([[2.3570225, 3.2998314, 1.4142135],
       [2.       , 0.       , 0.       ]], dtype=float32)>
```

Hope this helps!

**Comment by ageron** [comment]
I just realized that there's an error in this function: the mean of the embeddings should have been computed only on the word embeddings, excluding the padding tokens. Right now the function includes the padding tokens when computing the mean, which is incorrect.

So right now, for each sentence the function outputs S / (N + P) * sqrt(N) where S is the sum of all the word embeddings in the sentence, N is the number of non-padding tokens, and P is the number of padding tokens. The correct result should have been S / N * sqrt(N), which is equal to S / sqrt(N).

We can easily fix this issue by replacing `reduce_mean()` with `reduce_sum()`, and dividing the result by the square root of the number of words, instead of multiplying by this square root:

```python
def compute_mean_embedding(inputs):
    not_pad = tf.math.count_nonzero(inputs, axis=-1) 
    n_words = tf.math.count_nonzero(not_pad, axis=-1, keepdims=True)    
    sqrt_n_words = tf.math.sqrt(tf.cast(n_words, tf.float32))
    return tf.reduce_sum(inputs, axis=1) / sqrt_n_words
```

**Comment by ageron** [comment]
FYI, I just fixed this issue in commit https://github.com/ageron/handson-ml2/commit/0459e92a89e245a11202c582ef4c7531b29eead6

**Comment by kasrahabib** [comment]
Thanks!
It is much clear now!
